src/tmpfiles/tmpfiles.c

```diff
@@ -1288,45 +1287,113 @@ static int create_file(Item *i, const char *path) {
+                if (fd < 0)
+                        return log_error_errno(errno, "Failed to re-open file %s: %m", path);
 
-                r = -errno;
-                if (!i->argument && errno == EROFS && stat(path, &st) == 0 &&
-                    (i->type == CREATE_FILE || st.st_size == 0))
-                        goto check_mode;
+                if (fstat(fd, &st) < 0)
+                        return log_error_errno(errno, "stat(%s) failed: %m", path);
 
-                return log_error_errno(r, "Failed to create file %s: %m", path);
+                if (!S_ISREG(st.st_mode))
+                        return log_error_errno(EEXIST, "%s exists and is not a regular file.", path);
+
+                goto check_perms;
         }
 
+        log_debug("\"%s\" has been created.", path);
+
         if (i->argument) {
-                log_debug("%s to \"%s\".", i->type == CREATE_FILE ? "Appending" : "Writing", path);
+                log_debug("Writing to \"%s\".", path);
 
                 r = loop_write(fd, i->argument, strlen(i->argument), false);
                 if (r < 0)
                         return log_error_errno(r, "Failed to write file \"%s\": %m", path);
-        } else
-                log_debug("\"%s\" has been created.", path);
+        }
 
-        fd = safe_close(fd);
+check_perms:
+        return fd_set_perms(i, fd);
+}
 
- done:
-        if (stat(path, &st) < 0)
-                return log_error_errno(errno, "stat(%s) failed: %m", path);
+static int truncate_file(Item *i, const char *path) {
+        _cleanup_close_ int fd = -1;
+        int flags, r = 0;
+        struct stat st;
+        bool erofs = false;
+
+        assert(i);
+        assert(path);
+        assert(i->type == TRUNCATE_FILE);
+
+        /* We want to operate on regular file exclusively especially since
+         * O_TRUNC is unspecified if the file is neither a regular file nor a
+         * fifo nor a terminal device. Therefore we first open the file and make
+         * sure it's a regular one before truncating it. */
+
+        flags = O_CREAT|O_NOFOLLOW;
 
- check_mode:
-        if (!S_ISREG(st.st_mode)) {
-                log_error("%s is not a file.", path);
-                return -EEXIST;
+        RUN_WITH_UMASK(0000) {
+                mac_selinux_create_file_prepare(path, S_IFREG);
+                fd = open(path, flags|O_NDELAY|O_CLOEXEC|O_WRONLY|O_NOCTTY, i->mode);
+                mac_selinux_create_file_clear();
         }
 
-        r = path_set_perms(i, path);
-        if (r < 0)
-                return r;
+        if (fd < 0) {
+                if (errno != EROFS)
+                        return log_error_errno(errno, "Failed to open/create file %s: %m", path);
 
-        return 0;
+                /* On a read-only filesystem, we don't want to fail if the
+                 * target is already empty and the perms are set. So we still
+                 * proceed with the sanity checks and let fails the remaining
```

Comment:
"and let the remaining operations fail with EROFS"

Reply:
ok